wip-trampoline.md

```diff
@@ -0,0 +1,138 @@
+
+TODO: maybe this should be version 1 of the overall onion version namespace?
+
+TODO: maybe we can do something more clever with TLV? At the cost of a second
+onion parser (trampoline onion parser)? Use the top-level version in the parser
+to discriminate?
+
+TODO: version maybe redundant with the TLV type that will include this?
+But keeping it allows using this packet to be used as a top-level onion packet
+where we basically let each hop choose which channel to use to route to the
+next hop. Is this useful?
+
+The `trampoline_hops_data` field is a list of `trampoline_hop_payload`s.
+A `trampoline_hop_payload` is a structure that holds obfuscations of the next
+trampoline hop's address, transfer information, and its associated HMAC.
+
+It is `2*N*FRAME_SIZE` long and has the following structure (where `n1` and
+`n2` are the number of frames used by trampoline hops `1` and `2`):
+
+1. type: `trampoline_hops_data`
+2. data:
+   * [`n1*FRAME_SIZE`:`trampoline_hop_payload`]
+   * [`n2*FRAME_SIZE`:`trampoline_hop_payload`]
+   * ...
+   * `filler`
+
+Each `trampoline_hop_payload` has the following structure:
+
+1. type: `trampoline_hop_payload`
+2. data:
+   * [`1`:`num_frames_and_realm`]
+   * [`1`:`type`]
+   * [`1`:`length`]
+   * [`33`:`node_id`]
+   * [`8`:`amt_to_forward`]
+   * [`4`:`outgoing_cltv_value`]
+   * `optional TLV can be added here`
+   * `padding`
+   * [`32`:`HMAC`]
+
+For this specification, some restrictions apply:
+
+* `realm` MUST be `0x01` to indicate that the payload contains a TLV stream
+and `num_frames` MUST correctly reflect the number of frames used.
+* the TLV stream MUST contain the `type` `0x01` (`next_trampoline_hop`)
+* `next_trampoline_hop`'s length MUST be `33+8+4=45` bytes
+* the TLV stream MAY contain other optional types
+* `padding` MUST use only `0x00` bytes
+
+TODO: it's very sad to use 2 frames and have 50 bytes wasted per hop...
+But that can also be seen as an opportunity to add optional data to help routing ¯\_(ツ)_/¯
+Let's try to improve this though it would be great to be able to use a single
+frame per trampoline...
+
+### Nesting a Trampoline Onion in a v0 onion
+
+Trampoline onion packets combined with multi-frame v0 onion packets make it
+possible for nodes to route payments without a complete view of the network.
+
+The sending node (and each trampoline hop) uses TLV type `0x01` inside a v0
+onion packet to send a `trampoline_onion_packet` as last `hop_payload`:
+
+1. type: `onion_packet`
+2. data:
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`n1*FRAME_SIZE`:`hop_payload`]
+   * [`n2*FRAME_SIZE`:`hop_payload`]
+   * ...
+   * [`2*(N+1)*FRAME_SIZE`:`trampoline_onion_packet_payload`]
+   * `filler`
+   * [`32`:`hmac`]
+
+The structure of the last payload is:
+
+1. type: `trampoline_onion_packet_payload`
+2. data:
+   * [`1`:`realm`]
+   * [`1`:`type`]
+   * [`3`:`length`]
+   * [`2*N*FRAME_SIZE + 66`:`trampoline_onion_packet`]
+   * `optional TLV can be added here`
+   * `padding`
```

Comment:
Why not just define a TLV type and make it a completely normal TLV type-length-value pair? That way we could even give additional instructions to the trampoline, instead of having an optional TLV at the end.

Reply:
I'll investigate that, it's a bit of a trade-off between flexibility and space.
Thanks for the suggestion!